### PR TITLE
[EASI-3215] Allow multiple roles for collaborators

### PR DESCRIPTION
**Pull request by OddTomBrooks**
# EASI-3215

## Changes and Description

- Added flyway migration updating team_role to team_roles array
- Updated gql schema
- Updated resolvers
- Updated data model

## How to test this change

- Bring the systems up
- Create a plan collaborator using Postman
- Ensure the expected roles exist within Postman
- Update a plan collaborator's roles using Postman
- Delete the final Model Lead plan collaborator using Postman
  - This should throw an error

## PR Author Review Checklist

- [ ] Met the ticket's acceptance criteria, or will meet them in a subsequent PR.
- [ ] Added or updated tests for backend resolvers or other functions as needed.
- [ ] Added or updated client tests for new components, parent components, functions, or e2e tests as necessary.
- [ ] Updated the [Postman Collection](../MINT.postman_collection.json) if necessary.

## PR Reviewer Guidelines
- It's best to pull the branch locally and test it, rather than just looking at the code online!
- Check that all code is adequately covered by tests - if it isn't feel free to suggest the addition of tests.
- Always make comments, even if it's minor, or if you don't understand why something was done.

### CODE REVIEW

File: migrations/V107__Alter_Team_Role_To_Array.sql
Comment (ClayBenson94) [inline migrations/V107__Alter_Team_Role_To_Array.sql:22]: I don't think this function will continue to work as it was written [in this migration](https://github.com/CMSgov/mint-app/blob/8ad718c30bf3513130d9f27b6285869207797ba0/migrations/V8__Plan_Collaborator_Lead_Trigger.sql#L1-L18) - I think it'll need to be updated to handle the new column name and type

File: pkg/email/template_service_impl.go
Comment (ClayBenson94) [inline pkg/email/template_service_impl.go:75]: Guessing this will resolve itself when `main` is merged in, but seeing a lot of email template changes in this PR as-is

File: pkg/models/plan_collaborator.go
Comment (ClayBenson94) [inline pkg/models/plan_collaborator.go:51]: We already actually have a generic function for this!

https://github.com/CMSgov/mint-app/blob/79190855a4ab76afe7a2257c6c1054fb179ab2ba/pkg/models/model_helpers.go#L44-L51

File: pkg/shared/utilityCollections/search.go
Comment (ClayBenson94) [inline pkg/shared/utilityCollections/search.go:10]: Is this used anywhere? If not, we should delete it

We also often use [samber/lo](https://github.com/samber/lo) for functions like [this](https://github.com/samber/lo#contains)

**Review by ClayBenson94 — CHANGES_REQUESTED** [review]

File: pkg/email/template_service_impl.go
Comment (OddTomBrooks) [reply pkg/email/template_service_impl.go:75]: Yes. I branched off of the shared email branch on accident 😮‍💨 Surprised there are diffs though, expected it to not have diff against main.

File: pkg/shared/utilityCollections/search.go
Comment (OddTomBrooks) [reply pkg/shared/utilityCollections/search.go:10]: Good catch. I didn't like the process of looping through a collection but this attempt did not work. Will give samber a try next time.

File: migrations/V109__Alter_Team_Role_To_Array.sql
Comment (ClayBenson94) [inline migrations/V109__Alter_Team_Role_To_Array.sql:1]: image

It's probably worth adding a check here to ensure that you can't have multiple of the same role for one collaborator

File: migrations/V109__Alter_Team_Role_To_Array.sql
Comment (ClayBenson94) [inline migrations/V109__Alter_Team_Role_To_Array.sql:28]: image

Something isn't working correctly with this function locally. When I attempt to modify a collaborator (only collaborator on the plan, currently only has `["MODEL_LEAD"]` as their roles, I get this error even though the target roles also contain `["MODEL_LEAD"]`

However, I _don't_ get that error when the _only_ role I target is `"MODEL_LEAD"` - seems to be something with the `newRoles` array containing things that aren't `"MODEL_LEAD"` (seen below)

image

**Review by ClayBenson94 — CHANGES_REQUESTED** [review]

**Review by StevenWadeOddball — COMMENTED** [review]
Thanks Tom, I did some quick functional testing which worked as expected. I will continue with reviewing the code in the morning.

**Review by StevenWadeOddball — APPROVED** [review]
Thanks Tom, this looks good. I tested according the testing instructions and it worked as expected.

**Comment by OddTomBrooks** [comment]
This feature is complete from a backend perspective. I used git CLI to merge EASI-3217 into this branch which surprisingly caused GitHub to close PR #762. @garyjzhao please address the feedback from @patrickseguraoddball in this branch, as soon as we have front-end approval we will merge this to main.

**Review by patrickseguraoddball — CHANGES_REQUESTED** [review]
@garyjzhao Per the [figma](https://www.figma.com/file/M72DReE73SSXMN48WZfaKN/MINT---New-Model-Process-(Model-Intake)?type=design&node-id=6093-119354&mode=design&t=oJLtJ6UvhM3IB8rL-0) if there is only one collaborator/model lead, there should still be the ability to edit.  Once on the edit page, the ability to deselect a model lead should be disabled.  As well as the tag to remove a role - that X should be gone.  This is not a small change so we might want to loop back with UX and team to make sure we can invest in this.

**Comment by patrickseguraoddball** [comment]
Also the model lead always needs to be listed first [as shown here](https://www.figma.com/file/M72DReE73SSXMN48WZfaKN/MINT---New-Model-Process-(Model-Intake)?type=design&node-id=6169-124578&mode=design&t=rK6pbZmh26qPij4q-0)

The combobox "x" need to not render if there is only model lead in the selection [as shown here](https://www.figma.com/file/M72DReE73SSXMN48WZfaKN/MINT---New-Model-Process-(Model-Intake)?node-id=6093%3A119224&mode=dev)

**Review by patrickseguraoddball — CHANGES_REQUESTED** [review]
I'm running into an issue where I have a model lead already created.  Then I go to create another collaborator and select them as a model lead.  Then before I submit, I can't remove the tag for model lead, because it still thinks there's only one model lead

Screenshot 2023-10-26 at 12 44 06 PM

**Comment by patrickseguraoddball** [comment]
The app is also crashing on other multiselect components.  Here is the output

Screenshot 2023-10-26 at 12 45 09 PM

**Comment by patrickseguraoddball** [comment]
I notice that the clear indicator on the multiselect is never visible.  It should be visible if there are any selections and there is at least one model lead.  Here is what the multiselect looks like in dev

Screenshot 2023-10-26 at 12 47 47 PM

**Comment by patrickseguraoddball** [comment]
I notice that the clear indicator on the multiselect is not visible.  This seems related to the tag closing functionality listed above.  The clear indicator only appears if there are two model leads selected.  It should be visible if there are any selections and there is at least one model lead.  Here is what the multiselect looks like in dev.  

Screenshot 2023-10-26 at 12 47 47 PM

File: src/components/shared/MultiSelect/index.tsx
Comment (patrickseguraoddball) [inline src/components/shared/MultiSelect/index.tsx:260]: I think we don't want to default to ASC for sorting.  In almost all existing cases, we want to leave the tag in the order they are selected, so no sort necessary.

**Review by patrickseguraoddball — CHANGES_REQUESTED** [review]

**Review by patrickseguraoddball — CHANGES_REQUESTED** [review]
Sorry I missed this earlier but the alert isn't reflective of the current model lead.  This alert is reading as if the current collaborator is a model lead, but they arent

Screenshot 2023-10-26 at 3 22 32 PM

**Comment by patrickseguraoddball** [comment]
This should say `Roles`  instead of `Selected options`

Screenshot 2023-10-26 at 3 24 44 PM

https://www.figma.com/file/M72DReE73SSXMN48WZfaKN/MINT---New-Model-Process-(Model-Intake)?node-id=6093%3A119224&mode=dev

**Comment by patrickseguraoddball** [comment]
I think this whole page was intended to be revamped

https://www.figma.com/file/M72DReE73SSXMN48WZfaKN/MINT---New-Model-Process-(Model-Intake)?node-id=6199%3A10400&mode=dev

**Comment by patrickseguraoddball** [comment]
The roles should appear on the task list per Figma

https://www.figma.com/file/M72DReE73SSXMN48WZfaKN/MINT---New-Model-Process-(Model-Intake)?node-id=6199%3A10400&mode=dev

https://www.figma.com/file/M72DReE73SSXMN48WZfaKN/MINT---New-Model-Process-(Model-Intake)?node-id=6201%3A11454&mode=dev

**Comment by patrickseguraoddball** [comment]
This should now be `Manage model team members`

Screenshot 2023-10-26 at 3 30 43 PM

https://www.figma.com/file/M72DReE73SSXMN48WZfaKN/MINT---New-Model-Process-(Model-Intake)?node-id=6204%3A12404&mode=dev

**Review by patrickseguraoddball — CHANGES_REQUESTED** [review]
I realize we are still getting the non-model leads collaborators aren't sorted alphabetically by last name.  I think if we really want to get this functionality we need to do the both sorting outside of the table, and don't do any initial sort.

Screenshot 2023-11-01 at 1 05 43 PM

Additionally this sorting functionality should be replicated for both the task list and the readonly Teams page.  Can probably use the same sort util for all three

**Review by patrickseguraoddball — APPROVED** [review]
LG Thanks Gary!

**Pull request merged**
